swing-lib/src/main/java/org/triplea/swing/jpanel/JPanelBuilder.java

```diff
@@ -197,125 +78,40 @@ public JPanelBuilder horizontalAlignmentCenter() {
-    final Box box = Box.createHorizontalBox();
-    box.add(component);
-    box.add(Box.createHorizontalGlue());
-    return add(box);
-  }
-
-  public JPanelBuilder add(final Component component) {
-    Preconditions.checkNotNull(component);
-    panelComponents.add(
-        new PanelComponent(component, new PanelProperties(BorderLayoutPosition.DEFAULT)));
-    return this;
-  }
-
-  /** Adds a given component to the southern portion of a border layout. */
-  public JPanelBuilder addSouth(final JComponent child) {
-    layout = new BorderLayout();
-    Preconditions.checkNotNull(child);
-    panelComponents.add(new PanelComponent(child, new PanelProperties(BorderLayoutPosition.SOUTH)));
-    return this;
-  }
-
-  public JPanelBuilder addLabel(final String text) {
-    add(new JLabel(text));
-    return this;
-  }
-
-  /**
-   * use this when you want an empty space component that will take up extra space. For example,
-   * with a gridbag layout with 2 columns, if you have 2 components, the second will be stretched by
-   * default to fill all available space to the right. This right hand component would then resize
-   * with the window. If on the other hand a 3 column grid bag were used and the last element were a
-   * horizontal glue, then the 2nd component would then have a fixed size.
-   */
-  public JPanelBuilder addHorizontalGlue() {
-    add(Box.createHorizontalGlue());
+  public JPanelBuilder height(final int height) {
+    preferredHeight = height;
     return this;
   }
 
-  public JPanelBuilder addVerticalStrut(final int strutSize) {
-    add(Box.createVerticalStrut(strutSize));
-    return this;
+  public FlowLayoutBuilder add(final Component component) {
+    return flowLayout().add(component);
   }
 
-  public JPanelBuilder addHorizontalStrut(final int strutSize) {
-    add(Box.createHorizontalStrut(strutSize));
-    return this;
+  public FlowLayoutBuilder flowLayout() {
+    return new FlowLayoutBuilder(this);
   }
 
-  public JPanelBuilder borderLayout() {
-    layout = new BorderLayout();
-    return this;
+  public BorderLayoutBuilder borderLayout() {
+    return new BorderLayoutBuilder(this);
   }
 
-  public JPanelBuilder preferredHeight(final int height) {
-    preferredHeight = height;
-    return this;
+  public GridLayoutBuilder gridLayout(final int rows, final int columns) {
+    return new GridLayoutBuilder(this, rows, columns);
   }
 
-  /**
-   * BoxLayout needs a reference to the panel component that is using the layout, so we cannot
-   * create the layout until after we create the component. Thus we use a flag to create it, rather
-   * than creating and storing the layout manager directly as we do for the other layouts such as
-   * gridLayout.
-   */
-  private enum BoxLayoutType {
-    NONE,
-    HORIZONTAL,
-    VERTICAL
+  public BoxLayoutBuilder boxLayoutHorizontal() {
+    return new BoxLayoutBuilder(this, BoxLayoutBuilder.BoxLayoutOrientation.HORIZONTAL);
```

Comment:
Not exactly how I envisioned it but pretty similar.
I'd have expected the "specific" builders to extend the original builder, so that all common actions could stay there.
Not sure which approach is better, because my approach has the downside that the layout could be defined multiple times and this could only be checked at run-time, not at compile-time

Reply:
That's an interesting way to extend the common options to all builders.

I'd rather there be a stronger 'type-safe' way to do the builders, the classic builder pattern forces you to specify each required option, in order. I think to that extent, having an ordering requirement to specify the common options is okay. On the other hand, avoiding inheritance is significant, I could see additional issues coming up with inheritance (for example, method overriding could start making things very complicated)